src/main/java/de/tum/in/www1/artemis/service/connectors/AbstractContinuousIntegrationService.java

```diff
@@ -69,7 +75,61 @@ public Result createResultFromBuildResult(AbstractBuildResultNotificationDTO bui
      * @param result      the result for which the feedback should be added
      * @param buildResult The build result
      */
-    protected abstract void addFeedbackToResult(Result result, AbstractBuildResultNotificationDTO buildResult);
+    private void addFeedbackToResult(Result result, AbstractBuildResultNotificationDTO buildResult) {
+        final var jobs = buildResult.getBuildJobs();
+        final var programmingExercise = (ProgrammingExercise) result.getParticipation().getExercise();
+
+        // 1) add feedback for failed and passed test cases
+        addTestCaseFeedbacksToResult(result, jobs, programmingExercise);
+
+        // 2) process static code analysis feedback
+        addStaticCodeAnalysisFeedbackToResult(result, buildResult, programmingExercise);
+
+        // 3) process testwise coverage analysis report
+        addTestwiseCoverageReportToResult(result, buildResult, programmingExercise);
+
+        // Relevant feedback is negative
+        result.setHasFeedback(result.getFeedbacks().stream().anyMatch(feedback -> !feedback.isPositive()));
```

Comment:
This is not correct. With the [option for custom feedback on Jenkins systems](https://github.com/ls1intum/Artemis/pull/2881) a result can also have feedbacks if every test case is positive.

For Jenkis, the only line was this:
```
result.setHasFeedback(result.getFeedbacks().stream().anyMatch(feedback -> !feedback.isPositive() || feedback.getDetailText() != null));
```